File: en/java-springmvc/src/main/java/com/example/controller/InventoryController.java
package com.example.controller;

import org.springframework.web.bind.annotation.*;
import com.example.model.Inventory;
import com.example.dto.Result;
import com.example.dto.PageResult;
import jakarta.validation.Valid;
import java.math.BigDecimal;
import java.util.List;

@RestController
@RequestMapping("/api/inventory")
public class InventoryController {
    
    @GetMapping
    public Result<PageResult<Inventory>> listInventory(
            @RequestParam(required = false) Long productId,
            @RequestParam(required = false) String warehouse,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/{id}")
    public Result<Inventory> getInventory(@PathVariable Long id) {
        return Result.success(new Inventory());
    }
    
    @GetMapping("/product/{productId}")
    public Result<Inventory> getInventoryByProduct(@PathVariable Long productId) {
        return Result.success(new Inventory());
    }
    
    @PutMapping("/{id}")
    public Result<Inventory> updateInventory(@PathVariable Long id, @RequestParam Integer quantity) {
        return Result.success(new Inventory());
    }
    
    @PostMapping("/{id}/stock-in")
    public Result<Inventory> stockIn(
            @PathVariable Long id,
            @RequestParam Integer quantity,
            @RequestParam(required = false) BigDecimal costPrice,
            @RequestParam(required = false) String warehouse) {
        return Result.success(new Inventory());
    }
    
    @PostMapping("/{id}/stock-out")
    public Result<Inventory> stockOut(
            @PathVariable Long id,
            @RequestParam Integer quantity,
            @RequestParam(required = false) String warehouse) {
        return Result.success(new Inventory());
    }
    
    @PostMapping("/{id}/check")
    public Result<InventoryCheckVO> checkStock(@PathVariable Long id, @RequestParam Integer requiredQuantity) {
        return Result.success(new InventoryCheckVO());
    }
    
    @GetMapping("/{id}/history")
    public Result<PageResult<InventoryHistoryVO>> getHistory(
            @PathVariable Long id,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/low-stock")
    public Result<List<Inventory>> getLowStockProducts(@RequestParam(defaultValue = "10") Integer threshold) {
        return Result.success(List.of());
    }
    
    @GetMapping("/out-of-stock")
    public Result<List<Inventory>> getOutOfStockProducts() {
        return Result.success(List.of());
    }
    
    @PostMapping("/batch-stock-in")
    public Result<List<Inventory>> batchStockIn(@Valid @RequestBody List<StockInReq> reqs) {
        return Result.success(List.of());
    }
    
    @PostMapping("/batch-stock-out")
    public Result<List<Inventory>> batchStockOut(@Valid @RequestBody List<StockOutReq> reqs) {
        return Result.success(List.of());
    }
    
    @GetMapping("/statistics")
    public Result<InventoryStatisticsVO> getStatistics() {
        return Result.success(new InventoryStatisticsVO());
    }
    
    @GetMapping("/warehouse/{warehouse}")
    public Result<PageResult<Inventory>> getByWarehouse(
            @PathVariable String warehouse,
            @RequestParam(defaultValue = "1") Integer pageNum,
            @RequestParam(defaultValue = "10") Integer pageSize) {
        return Result.success(new PageResult<>());
    }
    
    @GetMapping("/warehouses")
    public Result<List<String>> getWarehouses() {
        return Result.success(List.of());
    }
    
    @PostMapping("/{id}/lock")
    public Result<Inventory> lockStock(@PathVariable Long id, @RequestParam Integer quantity) {
        return Result.success(new Inventory());
    }
    
    @PostMapping("/{id}/unlock")
    public Result<Inventory> unlockStock(@PathVariable Long id, @RequestParam Integer quantity) {
        return Result.success(new Inventory());
    }
    
    @GetMapping("/export")
    public Result<String> exportInventory() {
        return Result.success("export_file_url");
    }
    
    @PostMapping("/adjust")
    public Result<Inventory> adjustInventory(
            @PathVariable Long id,
            @RequestParam Integer quantity,
            @RequestParam String reason) {
        return Result.success(new Inventory());
    }
    
    @GetMapping("/alerts")
    public Result<List<InventoryAlertVO>> getAlerts() {
        return Result.success(List.of());
    }
}

class StockInReq {
    private Long productId;
    private Integer quantity;
    private BigDecimal costPrice;
    private String warehouse;
    
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public Integer getQuantity() { return quantity; }
    public void setQuantity(Integer quantity) { this.quantity = quantity; }
    public BigDecimal getCostPrice() { return costPrice; }
    public void setCostPrice(BigDecimal costPrice) { this.costPrice = costPrice; }
    public String getWarehouse() { return warehouse; }
    public void setWarehouse(String warehouse) { this.warehouse = warehouse; }
}

class StockOutReq {
    private Long productId;
    private Integer quantity;
    private String warehouse;
    
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public Integer getQuantity() { return quantity; }
    public void setQuantity(Integer quantity) { this.quantity = quantity; }
    public String getWarehouse() { return warehouse; }
    public void setWarehouse(String warehouse) { this.warehouse = warehouse; }
}

class InventoryCheckVO {
    private Long productId;
    private Integer availableQuantity;
    private Integer requiredQuantity;
    private Boolean isAvailable;
    
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public Integer getAvailableQuantity() { return availableQuantity; }
    public void setAvailableQuantity(Integer availableQuantity) { this.availableQuantity = availableQuantity; }
    public Integer getRequiredQuantity() { return requiredQuantity; }
    public void setRequiredQuantity(Integer requiredQuantity) { this.requiredQuantity = requiredQuantity; }
    public Boolean getIsAvailable() { return isAvailable; }
    public void setIsAvailable(Boolean isAvailable) { this.isAvailable = isAvailable; }
}

class InventoryHistoryVO {
    private Long id;
    private Long productId;
    private String action;
    private Integer quantity;
    private Integer beforeQuantity;
    private Integer afterQuantity;
    private String operator;
    private String createdAt;
    
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public String getAction() { return action; }
    public void setAction(String action) { this.action = action; }
    public Integer getQuantity() { return quantity; }
    public void setQuantity(Integer quantity) { this.quantity = quantity; }
    public Integer getBeforeQuantity() { return beforeQuantity; }
    public void setBeforeQuantity(Integer beforeQuantity) { this.beforeQuantity = beforeQuantity; }
    public Integer getAfterQuantity() { return afterQuantity; }
    public void setAfterQuantity(Integer afterQuantity) { this.afterQuantity = afterQuantity; }
    public String getOperator() { return operator; }
    public void setOperator(String operator) { this.operator = operator; }
    public String getCreatedAt() { return createdAt; }
    public void setCreatedAt(String createdAt) { this.createdAt = createdAt; }
}

class InventoryStatisticsVO {
    private Long totalProducts;
    private Long lowStockProducts;
    private Long outOfStockProducts;
    private Integer totalQuantity;
    private Integer totalValue;
    
    public Long getTotalProducts() { return totalProducts; }
    public void setTotalProducts(Long totalProducts) { this.totalProducts = totalProducts; }
    public Long getLowStockProducts() { return lowStockProducts; }
    public void setLowStockProducts(Long lowStockProducts) { this.lowStockProducts = lowStockProducts; }
    public Long getOutOfStockProducts() { return outOfStockProducts; }
    public void setOutOfStockProducts(Long outOfStockProducts) { this.outOfStockProducts = outOfStockProducts; }
    public Integer getTotalQuantity() { return totalQuantity; }
    public void setTotalQuantity(Integer totalQuantity) { this.totalQuantity = totalQuantity; }
    public Integer getTotalValue() { return totalValue; }
    public void setTotalValue(Integer totalValue) { this.totalValue = totalValue; }
}

class InventoryAlertVO {
    private Long productId;
    private String productName;
    private Integer currentQuantity;
    private Integer threshold;
    private String alertType;
    
    public Long getProductId() { return productId; }
    public void setProductId(Long productId) { this.productId = productId; }
    public String getProductName() { return productName; }
    public void setProductName(String productName) { this.productName = productName; }
    public Integer getCurrentQuantity() { return currentQuantity; }
    public void setCurrentQuantity(Integer currentQuantity) { this.currentQuantity = currentQuantity; }
    public Integer getThreshold() { return threshold; }
    public void setThreshold(Integer threshold) { this.threshold = threshold; }
    public String getAlertType() { return alertType; }
    public void setAlertType(String alertType) { this.alertType = alertType; }
}

File: en/kotlin-springmvc/src/main/kotlin/com/example/controller/InventoryController.kt
package com.example.controller

import com.example.dto.PageResult
import com.example.dto.Result
import com.example.model.Inventory
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/api/inventory")
class InventoryController {
    
    @GetMapping
    fun listInventory(
        @RequestParam(required = false) productId: Long?,
        @RequestParam(required = false) warehouse: String?,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Inventory>> {
        val result = PageResult<Inventory>()
        return Result.success(result)
    }
    
    @GetMapping("/{id}")
    fun getInventory(@PathVariable id: Long): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @GetMapping("/product/{productId}")
    fun getInventoryByProduct(@PathVariable productId: Long): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @PutMapping("/{id}")
    fun updateInventory(@PathVariable id: Long, @RequestParam quantity: Int): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @PostMapping("/{id}/stock-in")
    fun stockIn(
        @PathVariable id: Long,
        @RequestParam quantity: Int,
        @RequestParam(required = false) costPrice: BigDecimal?,
        @RequestParam(required = false) warehouse: String?
    ): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @PostMapping("/{id}/stock-out")
    fun stockOut(
        @PathVariable id: Long,
        @RequestParam quantity: Int,
        @RequestParam(required = false) warehouse: String?
    ): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @PostMapping("/{id}/check")
    fun checkStock(@PathVariable id: Long, @RequestParam requiredQuantity: Int): Result<InventoryCheckVO> {
        val check = InventoryCheckVO()
        return Result.success(check)
    }
    
    @GetMapping("/{id}/history")
    fun getHistory(
        @PathVariable id: Long,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<InventoryHistoryVO>> {
        val result = PageResult<InventoryHistoryVO>()
        return Result.success(result)
    }
    
    @GetMapping("/low-stock")
    fun getLowStockProducts(@RequestParam(defaultValue = "10") threshold: Int): Result<List<Inventory>> {
        val inventory = listOf<Inventory>()
        return Result.success(inventory)
    }
    
    @GetMapping("/out-of-stock")
    fun getOutOfStockProducts(): Result<List<Inventory>> {
        val inventory = listOf<Inventory>()
        return Result.success(inventory)
    }
    
    @PostMapping("/batch-stock-in")
    fun batchStockIn(@RequestBody reqs: List<StockInReq>): Result<List<Inventory>> {
        val inventory = listOf<Inventory>()
        return Result.success(inventory)
    }
    
    @PostMapping("/batch-stock-out")
    fun batchStockOut(@RequestBody reqs: List<StockOutReq>): Result<List<Inventory>> {
        val inventory = listOf<Inventory>()
        return Result.success(inventory)
    }
    
    @GetMapping("/statistics")
    fun getStatistics(): Result<InventoryStatisticsVO> {
        val statistics = InventoryStatisticsVO()
        return Result.success(statistics)
    }
    
    @GetMapping("/warehouse/{warehouse}")
    fun getByWarehouse(
        @PathVariable warehouse: String,
        @RequestParam(defaultValue = "1") pageNum: Int,
        @RequestParam(defaultValue = "10") pageSize: Int
    ): Result<PageResult<Inventory>> {
        val result = PageResult<Inventory>()
        return Result.success(result)
    }
    
    @GetMapping("/warehouses")
    fun getWarehouses(): Result<List<String>> {
        val warehouses = listOf<String>()
        return Result.success(warehouses)
    }
    
    @PostMapping("/{id}/lock")
    fun lockStock(@PathVariable id: Long, @RequestParam quantity: Int): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @PostMapping("/{id}/unlock")
    fun unlockStock(@PathVariable id: Long, @RequestParam quantity: Int): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @GetMapping("/export")
    fun exportInventory(): Result<String> {
        val fileUrl = ""
        return Result.success(fileUrl)
    }
    
    @PostMapping("/{id}/adjust")
    fun adjustInventory(
        @PathVariable id: Long,
        @RequestParam quantity: Int,
        @RequestParam(required = false) reason: String?
    ): Result<Inventory> {
        val inventory = Inventory()
        return Result.success(inventory)
    }
    
    @GetMapping("/alerts")
    fun getAlerts(): Result<List<InventoryAlertVO>> {
        val alerts = listOf<InventoryAlertVO>()
        return Result.success(alerts)
    }
}

data class StockInReq(
    val productId: Long?,
    val quantity: Int?,
    val costPrice: BigDecimal?,
    val warehouse: String?
)

data class StockOutReq(
    val productId: Long?,
    val quantity: Int?,
    val warehouse: String?
)

data class InventoryCheckVO(
    val productId: Long? = null,
    val availableQuantity: Int? = null,
    val requiredQuantity: Int? = null,
    val isAvailable: Boolean? = null
)

data class InventoryHistoryVO(
    val id: Long? = null,
    val productId: Long? = null,
    val action: String? = null,
    val quantity: Int? = null,
    val beforeQuantity: Int? = null,
    val afterQuantity: Int? = null,
    val operator: String? = null,
    val createdAt: String? = null
)

data class InventoryStatisticsVO(
    val totalProducts: Long? = null,
    val lowStockProducts: Long? = null,
    val outOfStockProducts: Long? = null,
    val totalQuantity: Int? = null,
    val totalValue: Int? = null
)

data class InventoryAlertVO(
    val productId: Long? = null,
    val productName: String? = null,
    val currentQuantity: Int? = null,
    val threshold: Int? = null,
    val alertType: String? = null
)
